### PR TITLE
simple plugin system for schnack

### PR DESCRIPTION
**Pull request by gka**
This PR refactors all of our currently supported authentication and notification methods into plugins.

The plugins are for now still part of this repo, but in theory they could already live in separate npm packages (and probably should be moved out soon as some might come with have hefty dependencies)

### Plugin configuration
Plugins can be defined and configured via the new `config.plugins` section. To enable the twitter authentication and slack notifications, one would set the plugins config to this:

```json
"plugins": {
    "auth-twitter": {
        "consumer_key": "xxxxx",
        "consumer_secret": "xxxxx"
    },
    "notify-slack": {
        "webhook_url": "xxxxx"
    }
}
```

### Plugin loading

In this PR plugins are loaded (= `require`d) using the [loadPlugins](https://github.com/schn4ck/schnack/blob/feature/plugins/src/helper.js#L94-L106) function. The function checks whether a plugin exists in one of two locations:
* the `src/plugins/${pluginId}` folder inside the Schnack repo, and
* a npm package called `schnack-plugin-${pluginId}`

As mentioned above, we should probably at some point move all "core" plugins into npm packages at some point to get them out of our package.dependencies.

The `loadPlugins` function is currently evoked at two occasions: in [src/auth.js](https://github.com/schn4ck/schnack/blob/feature/plugins/src/auth.js#L69) and in [src/notify.js](https://github.com/schn4ck/schnack/blob/feature/plugins/src/notify.js#L19) (formerly known as `src/push/index.js`).

There are two entry points that plugins can currently use: `auth` and `notify`.

### Plugin interface

As said, the plugins are being loaded at two locations, and each is only regarding plugins relevant to the part of Schnack where it's being loaded. In [src/auth.js](https://github.com/schn4ck/schnack/blob/feature/plugins/src/auth.js#L69) we are only invoking plugins that define the `auth()` hook, while in [src/notify.js](https://github.com/schn4ck/schnack/blob/feature/plugins/src/notify.js#L19) we only invoke plugins that define the `notify()` hook.

So while at this point, all auth and notify plugins are separate, in the future there could be plugins that handle both authentication and notification.

The hook functions `auth()` and `notify()` are really only passing some internal variables to the plugins (which cannot easily `require` them if they live in a separate npm package). This separation between the Schnack "core" and the plugins is meant to discourage ugly hacking and dependencies on internal helper functions etc. Essentially, whatever we're passing as arguments through these two hook functions become part of the new "public" API of Schnack and changes to those will potentially break backwards compatibility.

### Example plugins

I already ported all existing auth and notify methods into plugins so you see plenty examples. Here's how the [slack notification plugin](https://github.com/schn4ck/schnack/blob/feature/plugins/src/plugins/notify-slack/index.js#L1-L26) looks like. It uses the `events` object to observe new comments and then send http requests to the Slack API.

```js
const request = require('request');

module.exports = {
    notify({ events, config, page_url }) {
        events.on('new-comment', event => {
            try {
                const post_url = page_url.replace('%SLUG%', event.slug) + '#comment-' + event.id;
                const comment = event.comment
                    .split(/\n+/)
                    .map(s => (s ? `> _${s}_` : '>'))
                    .join('\n>\n');
                const text = `A <${post_url}|new comment> was posted by ${event.user.display_name ||
                    event.user.name} under *${event.slug}*:\n\n${comment}`;
                request({
                    url: config.webhook_url,
                    method: 'post',
                    json: true,
                    body: { text }
                });
            } catch (error) {
                console.error('Error sending slack notification:', error);
            }
        });
    }
};
```

### CODE REVIEW

File: create-schnack/index.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
+#! /usr/bin/env node
+/* eslint no-console: "off" */
+const fs = require('fs');
+const path = require('path');
+const { spawn } = require('child_process');
+const findUp = require('find-up');
```
Comment (sto3psl) [inline create-schnack/index.js]: This is not used.

Can be removed from `package.json` too.

Comment (gka) [reply create-schnack/index.js]: thx

File: create-schnack/index.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+   Please follow the setup instructions for schnack.
+
+   https://github.com/schn4ck/schnack#readme
+`);
+        process.exit(1);
+    }
+
+    const { plugins } = require(configPath);
+
+    const packages = Object.keys(plugins).map(id => `schnack-plugin-${id}`);
```
Comment (sto3psl) [inline create-schnack/index.js]: The script will crash if plugins are `undefined`

**Comment by joeshub** [comment]
just curious about the timeline for this, is this still under consideration?

**Comment by gka** [comment]
@joeshub apologies for the late reply. I will try to finish this release over the coming days.

**Pull request merged**